x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java

```diff
@@ -0,0 +1,123 @@
+
+    public static final ParseField RETRIEVER_FIELD = new ParseField("retriever");
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField RANK_WINDOW_SIZE_FIELD = new ParseField("rank_window_size");
+    public static final ParseField INFERENCE_ID_FIELD = new ParseField("inference_id");
+    public static final ParseField INFERENCE_TEXT_FIELD = new ParseField("inference_text");
+    public static final ParseField MIN_SCORE_FIELD = new ParseField("min_score");
+
+    public static final ObjectParser<TextSimilarityRankRetrieverBuilder, RetrieverParserContext> PARSER = new ObjectParser<>(
+        TextSimilarityRankBuilder.NAME,
+        TextSimilarityRankRetrieverBuilder::new
+    );
+
+    static {
+        PARSER.declareObject((r, v) -> r.retrieverBuilder = v, RetrieverBuilder::parseInnerRetrieverBuilder, RETRIEVER_FIELD);
+        PARSER.declareField((r, v) -> r.field = v, (p, c) -> p.text(), FIELD_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField(
+            (r, v) -> r.rankWindowSize = v,
+            (p, c) -> p.intValue(),
+            RANK_WINDOW_SIZE_FIELD,
+            ObjectParser.ValueType.INT_OR_NULL
+        );
+        PARSER.declareField((r, v) -> r.inferenceId = v, (p, c) -> p.text(), INFERENCE_ID_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.inferenceText = v, (p, c) -> p.text(), INFERENCE_TEXT_FIELD, ObjectParser.ValueType.STRING);
+        PARSER.declareField((r, v) -> r.minScore = v, (p, c) -> p.floatValue(), MIN_SCORE_FIELD, ObjectParser.ValueType.FLOAT_OR_NULL);
+
+        RetrieverBuilder.declareBaseParserFields(TextSimilarityRankBuilder.NAME, PARSER);
+    }
+
+    public static TextSimilarityRankRetrieverBuilder fromXContent(XContentParser parser, RetrieverParserContext context)
+        throws IOException {
+        if (context.clusterSupportsFeature(TEXT_SIMILARITY_RERANKER_RETRIEVER_SUPPORTED) == false) {
+            throw new ParsingException(parser.getTokenLocation(), "unknown retriever [" + TextSimilarityRankBuilder.NAME + "]");
+        }
+        if (TextSimilarityRankBuilder.TEXT_SIMILARITY_RERANKER_FEATURE.check(XPackPlugin.getSharedLicenseState()) == false) {
+            throw LicenseUtils.newComplianceException(TextSimilarityRankBuilder.NAME);
+        }
+        return PARSER.apply(parser, context);
+    }
+
+    private RetrieverBuilder retrieverBuilder;
+    private String field;
+    private int rankWindowSize;
+    private String inferenceId;
+    private String inferenceText;
+    private Float minScore;
+
+    public TextSimilarityRankRetrieverBuilder() {
+        this.rankWindowSize = DEFAULT_RANK_WINDOW_SIZE;
+    }
+
+    @Override
+    public void extractToSearchSourceBuilder(SearchSourceBuilder searchSourceBuilder, boolean compoundUsed) {
+        retrieverBuilder.extractToSearchSourceBuilder(searchSourceBuilder, compoundUsed);
+
+        if (searchSourceBuilder.rankBuilder() == null) {
```

Comment:
@pmpailis Added code to throw exception in this case: https://github.com/elastic/elasticsearch/pull/109813/commits/cddc7cfe2fb7e4de002b39f9f9c73922f6f60f35.

Note it's not possible to reach this state because retriever nesting is limited to 2 levels deep, but it's good to have this safety net.